In [ ]:
Breast Cancer Detection using Logistic Regression

In [ ]:
#Importing necessary libraries
import numpy as np
import sklearn.datasets

In [ ]:
#Getting the dataset
breast_cancer=sklearn.datasets.load_breast_cancer()

In [ ]:
print(breast_cancer)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [ ]:
#Data array:It contains all the observed values from the test
#Target array:It contains zero or one ;0 for malignant ;1 for benign
#0 and 1 are called labels

In [ ]:
#Loading data into X and target into Y

In [ ]:
X = breast_cancer.data
Y =breast_cancer.target

In [ ]:
print(X)
print(Y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

In [ ]:
#Checking the shape of X and Y
print(X.shape,Y.shape)

(569, 30) (569,)


So we have 569 training examples and we are using 30 features (569,30)

In [ ]:
#///////////////////////////////////////////////////

***Importing data to pandas Data Frame***

In [ ]:
import pandas as pd

In [ ]:
data=pd.DataFrame(breast_cancer.data,columns=breast_cancer.feature_names)

In [ ]:
data['class'] = breast_cancer.target

In [ ]:
data.head()

mean radius  mean texture  ...  worst fractal dimension  class
0        17.99         10.38  ...                  0.11890      0
1        20.57         17.77  ...                  0.08902      0
2        19.69         21.25  ...                  0.08758      0
3        11.42         20.38  ...                  0.17300      0
4        20.29         14.34  ...                  0.07678      0

[5 rows x 31 columns]

In [ ]:
#Statistics of the data

In [ ]:
data.describe()

mean radius  mean texture  ...  worst fractal dimension       class
count   569.000000    569.000000  ...               569.000000  569.000000
mean     14.127292     19.289649  ...                 0.083946    0.627417
std       3.524049      4.301036  ...                 0.018061    0.483918
min       6.981000      9.710000  ...                 0.055040    0.000000
25%      11.700000     16.170000  ...                 0.071460    0.000000
50%      13.370000     18.840000  ...                 0.080040    1.000000
75%      15.780000     21.800000  ...                 0.092080    1.000000
max      28.110000     39.280000  ...                 0.207500    1.000000

[8 rows x 31 columns]

In [ ]:
print(data['class'].value_counts())

1    357
0    212
Name: class, dtype: int64


In [ ]:
print(breast_cancer.target_names)

['malignant' 'benign']


In [ ]:
data.groupby('class').mean()

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
class                             ...                                         
0        17.462830     21.604906  ...        0.323468                 0.091530
1        12.146524     17.914762  ...        0.270246                 0.079442

[2 rows x 30 columns]

0 -Malignant
1 -benign

Train and test split

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.1)

Using 10% of data for testing and 90 % of the data for training

In [ ]:
print(Y.shape,Y_train.shape,Y_test.shape)

(569,) (512,) (57,)


We need the mean of original data ,train data and test data to be almost equal.Lets see how it looks 

In [ ]:
print(Y.mean(),Y_train.mean(),Y_test.mean())

0.6274165202108963 0.623046875 0.6666666666666666


In [ ]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.1,stratify=Y)

In [ ]:
print(Y.mean(),Y_train.mean(),Y_test.mean())

0.6274165202108963 0.626953125 0.631578947368421


In [ ]:
#Random state is used to reproduce the code so that the data is split in the same way

In [ ]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.1,stratify=Y,random_state=2)

In [ ]:
#test_size ---->To specify the percentage of test data needed
#stratify ----->for correct distribution as of the original data
#random_state ----->specify split of data:each value of random_splits the data differently

In [ ]:
#There are many features in X train hence it is ok if it contains variations in mean

In [ ]:
print(X.mean(),X_train.mean(),X_test.mean())

61.890712339519624 62.369174136692706 57.592950231578946


**LOGISTIC REGRESSION**

In [ ]:
#Importing logistic regression from sklearn
from sklearn.linear_model import LogisticRegression

In [ ]:
#Loading the logistic regression model to the variable 'classifier'
classifier=LogisticRegression()

In [ ]:
classifier.fit?

In [ ]:
#Training the model on training data
classifier.fit(X_train,Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

***Evaluation of the model***

In [ ]:
#import accuracy_score
from sklearn.metrics import accuracy_score

In [ ]:
prediction_on_training_data = classifier.predict(X_train)
accuracy_on_training_data = accuracy_score(Y_train,prediction_on_training_data)

In [ ]:
print("Accuracy on training data :",accuracy_on_training_data)

Accuracy on training data : 0.947265625


In [ ]:
#Finding the accuracy on test data

In [ ]:
#Prediction on test data
prediction_on_test_data=classifier.predict(X_test)
accuracy_on_testing_data=accuracy_score(Y_test,prediction_on_test_data)

In [ ]:
print("Accuracy on the test data : ",accuracy_on_testing_data)

Accuracy on the test data :  0.9824561403508771


Detecting whether the patient has Breast cancer in benign or Malignant stage

In [ ]:
input_data=(17.99,10.38,122.8,1001,0.1184,0.2776,0.3001,0.1471,0.2419,0.07871,1.095,0.9053,8.589,153.4,0.006399,0.04904,0.05373,0.01587,0.03003,0.006193,25.38,17.33,184.6,2019,0.1622,0.6656,0.7119,0.2654,0.4601,0.1189)

In [ ]:
#change the input_data to numpy array to make prediction

In [ ]:
input_data_as_numpy_array=np.asarray(input_data)
print(input_data_as_numpy_array)

[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
 1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
 6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
 1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
 4.601e-01 1.189e-01]


In [ ]:
#Reshape the array as we are predicting the output for one instance

In [ ]:
input_data_reshaped =input_data_as_numpy_array.reshape(1,-1)

In [ ]:
#Prediction
prediction=classifier.predict(input_data_reshaped)
print(prediction)
#Returns a list with element [0] if Malignant ;Returns a list with element[1] ,if benign
if prediction[0]==0:
  print("The Breast cancer in Malignant")
else :
  print("The Breast cancer in Benign")

[0]
The Breast cancer in Malignant
